docs/proposals/007-variadic-generics.md

```diff
@@ -0,0 +1,662 @@
+requirement that these type packs will have the same number of elements. For example:
+```
+ // implicitly requiring T and U to have same number of elements.
+void g<each T, each U>(expand Pair<each T, each U> pairs) {...}
+
+void user()
+{
+    // We will match (int, float) to `T`, and (uint16_t, half) to `U`:
+    g<int, float, uint16_t, half>(
+        Pair<int, uint16_t>(1, 2),
+        Pair<float, half>(1.0f, 2.0h) );
+}
+```
+
+In the example above, the type `expand Pair<each T, each U>` defines a pack of types where each element in the pack is formed by
+replacing `each T` and `each U` in the __pattern type__ `Pair<each T, each U>` with the corresponding elements in type pack `T` and `U`.
+Because the pattern type `Pair<each T, each U>` references two different type pack parameters `T` and `U`, we require that `T` and `U`
+has the same number of types, this allows us to resolve `g<int, float, uint16_t, half>` by evenly dividing the the argument list
+into two parts, such that `T = (int, float)` and `U = (uint16_t, half)`. With that, `expand Pair<each T, each U>` is then substituted
+into a type pack `(Pair<int, uint16_t>, Pair<float, half>)`.
+
+Generic type pack parameters can have type constraints:
+
+```
+void f<each T : IFloat>(expand each T v) {}
+```
+
+This means that every type in the type pack `T` must conform to the interface `IFloat`.
+You can use any expression inside `expand` when it is used on values:
+```
+interface IGetValue
+{
+    int getValue();
+}
+
+void print(each T)(expand each T) {...}
+
+void f<each T : IFloat>(expand each T v)
+{
+    print(expand (each v).getValue());
+}
+```
+
+Here, `expand (each v).getValue()` will expand the pattern expression `(each v).getValue()` into a pack of values. The result of this `expand` expression
+is a pack of values where each element of the pack is computed by substituting `each v` in the pattern expression with each element in `v`. The resulting
+pack of `int` values is then passed to `print` function that also takes a pack of values.
+
+For now, we require that all variadic generic type packs to appear in the end of a parameter list, after any ordinary parameters. This means that the following
+definitions are invalid:
+
+```
+void f<each T, U>() {} // Error, ordinary parameeter `U` after type pack.
+void g<each T, U = int>() {} // Error, ordinary parameter after type pack.
+void k<each T, let i : int>() {} // Error.
+void h<U = int, each T>() {} // OK.
+```
+
+Addtionally, we establish these restrictions on how `expand` and `each` maybe used:
+- The pattern type of an `expand` type expression must capture at least one generic type pack parameter in an `each` expression.
+- The type expression after `each` must refer to a generic type pack parameter, and the `each` expression can only appear inside an `expand` expression.
+
+These rules means that expressions like `expand int`, or `each T` on its own are invalid expressions.
```

Comment:
Maybe I understand wrong, but looking at above exmaple:
`print(expand (each v).getValue());`

This seems will become `print(expand int);` is that OK?

Or, is this rule only for the declaration?

Reply:
That is OK, because the expansion process is:
```
print (expand (each v).getValue())

==>

print( MakeTuple(v[0].getValue(), v[1].getValue(), ...., v[n].getValue() ) );
```

Reply:
Regardless of types or values, `expand X` just means duplicate `X` n times, and replace `each something` with `something[i]`. The number of times to duplicate is gathered by looking at the type/value packes captured in `X`, and not the return type of `X`.

Reply:
This makes lots of more sense to me, thanks.